src/flamenco/log_collector/fd_log_collector.h

```diff
@@ -0,0 +1,526 @@
+#ifndef HEADER_fd_src_flamenco_log_collector_fd_log_collector_h
+#define HEADER_fd_src_flamenco_log_collector_fd_log_collector_h
+
+#include "fd_log_collector_base.h"
+#include "../runtime/context/fd_exec_instr_ctx.h"
+#include "../runtime/context/fd_exec_txn_ctx.h"
+#include "../../ballet/base58/fd_base58.h"
+#include "../../ballet/base64/fd_base64.h"
+#include <stdio.h>
+#include <stdarg.h>
+
+/* Log collector + stable log implementations.
+   https://github.com/anza-xyz/agave/blob/v2.0.6/program-runtime/src/log_collector.rs
+   https://github.com/anza-xyz/agave/blob/v2.0.6/program-runtime/src/stable_log.rs */
+
+/* INTERNALS
+   Internal functions, don't use directly. */
+
+#define FD_EXEC_LITERAL(STR) ("" STR), (sizeof(STR)-1)
+
+/* _fd_log_collector_push pushes a log (internal, don't use directly).
+
+   This function stores a log msg of size msg_sz, serialized as protobuf.
+   For the high-level functionality, see fd_log_collector_msg().
+
+   Internally, each log msg is serialized as a 1 byte tag + 1 or 2 bytes
+   msg_sz (variable int < 32768) + msg_sz bytes of the actual msg.
+
+     |  tag   |    msg_sz    |     msg      |
+     | 1-byte | 1-or-2 bytes | msg_sz bytes |
+
+   The advantage of this representation is that when we have to store
+   txn metadata in blockstore, we don't have to do any conversion for logs,
+   just copy the entire buffer. */
+static inline void
+_fd_log_collector_push( fd_log_collector_t * log,
+                        char const *         msg,
+                        ulong                msg_sz ) {
+  uchar * buf   = log->buf;
+  ulong   buf_sz = log->buf_sz;
+
+  /* Store tag + msg_sz */
+  ulong needs_2b  = (msg_sz>0x7F);
+  buf[ buf_sz   ] = FD_LOG_COLLECTOR_PROTO_TAG;
+  buf[ buf_sz+1 ] = (uchar)( (msg_sz&0x7F) | (needs_2b<<7) );
+  buf[ buf_sz+2 ] = (uchar)( (msg_sz>>7) & 0x7F ); /* This gets overwritten if 0 */
+
+  /* Copy msg and update total buf_sz */
+  ulong msg_start = buf_sz + 2 + needs_2b;
+  fd_memcpy( buf + msg_start, msg, msg_sz );
+  log->buf_sz = (ushort)( msg_start + msg_sz );
+}
+
+/* _fd_log_collector_debug prints all logs (internal, don't use directly). */
+static inline void
+_fd_log_collector_debug( fd_log_collector_t const * log );
+
```

Comment:
Learning your coding style here... what does prefixing with _ represent?

Reply:
Usually private methods, but I see the confusion - renamed it to private so it's more clean